subprojects/internal-integ-testing/src/main/groovy/org/gradle/integtests/fixtures/logging/comparison/ExhaustiveLinesSearcher.java

```diff
@@ -0,0 +1,183 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.integtests.fixtures.logging.comparison;
+
+import org.apache.commons.lang3.StringUtils;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.Set;
+
+public class ExhaustiveLinesSearcher {
+    private boolean matchBlankLines = false;
+    private boolean showLessLikelyMatches = false;
+    private boolean useUnifiedDiff = false;
+
+    /**
+     * Instructs empty lines to be considered matches when exhaustively building the list of potential matches.
+     * <p>
+     * Enabling this option could really explode the number of potential matches for common output cases
+     * where blank lines exist in both the expected and actual lines.
+     *
+     * @return {@code this)
+     */
+    public ExhaustiveLinesSearcher matchesBlankLines() {
+        matchBlankLines = true;
+        return this;
+    }
+
+    /**
+     * Show potential matches which don't match as many expected lines as the best match.
+     *
+     * @return {@code this)
+     */
+    public ExhaustiveLinesSearcher showLessLikelyMatches() {
+        showLessLikelyMatches = true;
+        return this;
+    }
+
+    /**
+     * Display the most likely potential match according to the git diff algorithm.
+     * <p>
+     * Enabling this option will only show a single match, regardless of how many potential matches
+     * containing the same number of matching lines are found.
+     *
+     * @return {@code this}
+     */
+    public ExhaustiveLinesSearcher useUnifiedDiff() {
+        useUnifiedDiff = true;
+        return this;
+    }
+
+    public void assertLinesContainedIn(List<String> expectedLines, List<String> actualLines) throws LineSearchFailures.AbstractLineListComparisonFailure {
+        if (expectedLines.isEmpty() && actualLines.isEmpty()) {
+            return;
+        }
+
+        if (actualLines.size() < expectedLines.size()) {
+            LineSearchFailures.insufficientSize(expectedLines, actualLines);
+        }
+
+        if (!findFirstCompleteMatch(expectedLines, actualLines).isPresent()) {
+            exhaustiveSearch(expectedLines, actualLines);
+        }
+    }
+
+    public void assertSameLines(List<String> expectedLines, List<String> actualLines) {
+        if (expectedLines.isEmpty() && actualLines.isEmpty()) {
+            return;
+        }
+
+        if (expectedLines.size() != actualLines.size()) {
+            LineSearchFailures.differentSizes(expectedLines, actualLines);
+        }
+
+        if (!findFirstCompleteMatch(expectedLines, actualLines).orElse(-1).equals(0)) {
```

Comment:
Why `.equals(0)` instead of `== 0`?

Reply:
To avoid checking reference equality, of course, in case they happen to be two different zero instances.  Good catch, I'll change it.